cdc/sink/producer/kafka/kafka.go

```diff
@@ -395,80 +400,127 @@ func (k *kafkaSaramaProducer) run(ctx context.Context) error {
 	}
 }
 
-// kafkaTopicPreProcess gets partition number from existing topic, if topic doesn't
-// exit, creates it automatically.
-func kafkaTopicPreProcess(topic, address string, config *Config, cfg *sarama.Config) (int32, error) {
-	admin, err := sarama.NewClusterAdmin(strings.Split(address, ","), cfg)
+func topicPreProcess(topic string, config *Config, saramaConfig *sarama.Config) error {
+	// FIXME: find a way to remove this failpoint for workload the unit test
+	failpoint.Inject("workaround4Test", func() {
+		failpoint.Return(nil)
+	})
+	admin, err := sarama.NewClusterAdmin(config.BrokerEndpoints, saramaConfig)
 	if err != nil {
-		return 0, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+		return err
 	}
 	defer func() {
-		err := admin.Close()
-		if err != nil {
-			log.Warn("close admin client failed", zap.Error(err))
+		if err := admin.Close(); err != nil {
+			log.Warn("close kafka cluster admin failed", zap.Error(err))
 		}
 	}()
+
 	topics, err := admin.ListTopics()
 	if err != nil {
-		return 0, cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
-	}
-	partitionNum := config.PartitionNum
-	topicDetail, exist := topics[topic]
-	if exist {
-		log.Info("get partition number of topic", zap.String("topic", topic), zap.Int32("partition_num", topicDetail.NumPartitions))
-		if partitionNum == 0 {
-			partitionNum = topicDetail.NumPartitions
-		} else if partitionNum < topicDetail.NumPartitions {
-			log.Warn("partition number assigned in sink-uri is less than that of topic", zap.Int32("topic partition num", topicDetail.NumPartitions))
-		} else if partitionNum > topicDetail.NumPartitions {
-			return 0, cerror.ErrKafkaInvalidPartitionNum.GenWithStack(
-				"partition number(%d) assigned in sink-uri is more than that of topic(%d)", partitionNum, topicDetail.NumPartitions)
+		return err
+	}
+
+	var realPartitionCount int32 = 0
+	info, ok := topics[topic]
+	if ok {
+		realPartitionCount = info.NumPartitions
+	}
+
+	// if user specify cdc to create the topic, we must make sure
+	// that `partition-num` and `replication-factor` is given
+	if config.AutoCreate {
+		// if the specified topic name already exist, we should make sure that
+		// topic's `max.message.bytes` is not less than given `max-message-size`
+		if ok {
+			if a, ok := info.ConfigEntries["max.message.bytes"]; ok {
+				topicMaxMessageBytes, err := strconv.Atoi(*a)
+				if err != nil {
+					return cerror.WrapError(cerror.ErrKafkaNewSaramaProducer, err)
+				}
+				if topicMaxMessageBytes < config.MaxMessageBytes {
```

Comment:
Do we also need to check max message bytes when topic exists and `AutoCreate` is false?